xdsl/dialects/fsm.py

```diff
@@ -0,0 +1,379 @@
+    arg_names = opt_attr_def(ArrayAttr[StringAttr])
+    res_names = opt_attr_def(ArrayAttr[StringAttr])
+
+    traits = frozenset([NoTerminator(), SymbolTable()])
+
+    def __init__(
+        self,
+        sym_name: str,
+        initial_state: str,
+        function_type: FunctionType | tuple[Sequence[Attribute], Sequence[Attribute]],
+        arg_attrs: ArrayAttr[DictionaryAttr] | None,
+        res_attrs: ArrayAttr[DictionaryAttr] | None,
+        arg_names: ArrayAttr[StringAttr] | None,
+        res_names: ArrayAttr[StringAttr] | None,
+        body: Region | type[Region.DEFAULT] = Region.DEFAULT,
+    ):
+        if isinstance(function_type, tuple):
+            inputs, outputs = function_type
+            function_type = FunctionType.from_lists(inputs, outputs)
+        attributes: dict[str, Attribute | None] = {
+            "sym_name": StringAttr(sym_name),
+            "initialState": StringAttr(initial_state),
+            "function_type": function_type,
+            "arg_attrs": arg_attrs,
+            "res_attrs": res_attrs,
+            "arg_names": arg_names,
+            "res_names": res_names,
+        }
+
+        if not isinstance(body, Region):
+            body = Region(Block())
+
+        super().__init__(attributes=attributes, regions=[body])
+
+    def verify_(self):
+        if not SymbolTable.lookup_symbol(self, self.initialState):
+            raise VerifyException("Can not find initial state")
+        # if arg_names is given, arg_attrs must be given and consistent
+        if (self.arg_attrs is None) ^ (self.arg_names is None):
+            raise VerifyException("arg attrs must be consistent with names")
+        # if res_names is given, res_attrs must be given and consistent
+        if (self.res_attrs is None) ^ (self.res_names is None):
+            raise VerifyException("res attrs must be consistent with names")
+        if self.arg_attrs is not None and self.arg_names is not None:
+            if len(self.arg_attrs) != len(self.arg_names):
+                raise VerifyException(
+                    "The number of arg_attrs and arg_names should be the same"
+                )
+        if self.res_attrs is not None and self.res_names is not None:
+            if len(self.res_attrs) != len(self.res_names):
+                raise VerifyException(
+                    "The number of res_attrs and res_names should be the same"
+                )
+
+
+@irdl_op_definition
+class OutputOp(IRDLOperation):
+    """
+    Represents the outputs of a machine under a specific state. The
+    types of `$operands` should be consistent with the output types of the state
+    machine
+    """
+
+    name = "fsm.output"
+
+    operand = var_operand_def(AnyAttr())
+
+    traits = frozenset([IsTerminator()])
+
+    def __init__(
+        self,
+        operand: Sequence[SSAValue | Operation],
+    ):
+        super().__init__(
+            operands=[operand],
+        )
+
+    def verify_(self):
+        parent = self.parent_op()
```

Comment:
I would add an `assert(isinstance(parent, StateOp))` here, since the operation has the `HasParent` trait (which is checked before the `verify_` method.